### DISCUSSION TITLE
프론트엔드 개발에서 선언적 프로그래밍 패러다임은 기본적인 개념이 되었습니다.

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#66] [프론트 기술 질문 공간]
1. “선언적 프로그래밍”이라고 하는 개념에 대해 소개해주세요.
    1. 선언적인 코드는 어떤 장점을 가지고 있나요?
    2. 최근에 선언적으로 작성하려고 한 코드 조각을 소개해주세요. 어떤 점이 좋아졌나요?

**Comment by michoball** [comment]
선언적 프로그래밍은 
프로그램이 어떤 식으로 동작해야하는지 일일히 명시적으로 코드를 짜는 것이 아니라 원하는 결과를 명시하는 프로그래밍입니다. 
함수가 무엇을 해야하는지를 중점적으로 다루며 그안에 세부사항들은 추상화되어있습니다. 

  1.  코드의 가독성을 향상시키고 재사용성이나 유지보수성을 높입니다.  

  2.  url search param 값을 다루는 커스텀 훅을 만든 적이 있습니다. 그 중 기존에 있는 searchParam에 새로운 param를 덧붙여주는  setSearchParams라는 함수입니다. 
  이 코드를 사용하는 쪽에서는 키와 값만을 객체로 { newParam: 1 } 과 같이 사용하기만 하면 됩니다.  내부적인 동작은 몰라도 됩니다. 
 ```typescript
  const setSearchParams = useCallback(
    (setParams: SearchParams) => {
      Object.keys(setParams).forEach((key) => {
        urlTerm.set(key, setParams[key]);
        serUrlTerm(urlTerm);
      });
    },
    [urlTerm, serUrlTerm]
  );
```

**Comment by javascriptbiggosoo** [comment]
1. 코드 작성자에게 재사용성을 높여주면서 동작의 세부적인 로직이 드러나지않고 결과를 한눈에 알아볼 수 있다는 장점이 있습니다.
2. 리액트 개발자라면 누구나 선언형 방식의 코딩의 이점을 누리고 있습니다.
```tsx
        <Items>
          <Item>
            <Link to="/" style={{ fontWeight: HomePage ? "bold" : "normal" }}>
              Home
            </Link>
          </Item>
          <Item>
            <Link to="/tv" style={{ fontWeight: TVPage ? "bold" : "normal" }}>
              TV Shows
            </Link>
          </Item>
        </Items>
```
위 코드를 jsx가 아닌 javascript로 구현한다면
모든 컴포넌트 혹은 엘리먼트를 직접 생성(`createElement`, `innerHTML`) 해준 후
삽입할 위치를 찾아서(`querySelector`) 추가(`append`)해주고
어떤 텍스트(`innerText`)가 들어가는지, 어떤 이벤트핸들러(`addEventHandler`)가 들어가는지조차 한땀한땀 작성해줘야 합니다.

**Comment by BaekYuri** [comment]
“선언적 프로그래밍”이라고 하는 개념에 대해 소개해주세요.
-> 원하는 결과를 어떻게 얻을 것인지 명시하지 않고, 원하는 결과가 무엇인지 명시하여 프로그램을 작성하는 프로그래밍입니다.
1. 선언적인 코드는 어떤 장점을 가지고 있나요?
-> 간결하게 작성할 수 있고, "선언"이므로 가독성이 좋습니다.
2. 최근에 선언적으로 작성하려고 한 코드 조각을 소개해주세요. 어떤 점이 좋아졌나요?
-> 오늘의 거래내역만 필터링하는 코드입니다. 직관적으로 알 수 있고, 이 코드가 어떻게 동작하는지는 알 필요가 없습니다.
```javascript
initTrade.filter(trade => trade.transaction_time.includes(today))
```

**Comment by WooGie911** [comment]
선언적 프로그래밍이란 기계의 작동 모델을 기반으로 프로그래밍 하는 것이 아닌, 개발자의 생각 모델을 기반으로 하는 개발입니다.

선언적인 코드는 어떤 장점을 가지고 있나요?
->
1. 가독성이 좋다. 코드의 의도를 명확하게 표현할 수 있기 때문이다.
2. 재사용성이 좋다. 코드를 모듈화하고 재사용 가능한 단위로 분리하기 쉽게 만든다.
3. 유지보수 하기 좋다. 작업의 세부 사항을 추상화하고 숨김으로써 코드의 유지보수를 용이하게 만들기 때문이다.
4. 테스트 하기 좋다. 의존성을 명시적으로 표현하고 숨기기 때문에 테스트하기 쉽다.
5. 예측 가능하다. 명시적인 규칙과 제약 조건을 따르도록 유도하기 때문에 오류를 방지하고 예측 가능한 동작을 유지하는 데 도움이 된다.
